### ISSUE TITLE
Ryukishi sprites

### ISSUE THREAD

**Issue by enumag** [07th-mod/higurashi-patch-compiler#54]
Currently our patch supports the PS3 and MG sprites but we removed the support for the original Ryukishi sprites. In this issue I'll write down the steps we need to take to add them back.

1. JSON file with sprite aliases
----

Both MG and Ryukishi sprites have fewer facial expressions than PS3. For MG sprites we solved it by addind certain sprites as multiple copies with different names which increases the patch size unnecessarily.

A better way would be to add a JSON file with a map of sprite aliases.

For Ryukishi sprites this would enable us to use the sprite names from vanilla and just link them to PS3 sprites using the json.

2. Fallback for sprites when LS is not active
----

I noticed that most cases of sprite duplication in CGAlt are cases like `me1a_akuwarai_a1_1` and `me1a_akuwarai_a1_2` being the same sprite. This is intentional as the sprite is different for PS3 sprite with LS disabled. For CGAlt there are also some cases where the sprites are different such as `oisi1_5_0` and `oisi1_5_2` in Watanagashi.

Solution: For CGAlt (and for Ryikishi sprites) if `sprite_X` does not exist try `sprite_(X-1)` and `sprite_(X-2)` (if X=2) as fallbacks before falling back to PS3 `sprite_X`.

3. Write the list of aliases
----

Next we need the linking for vanilla names (Ryukishi sprites) to our names (PS3 sprites). This will need to be done manually and will take some time because there are several hundreds of sprites for each arc. Of course we can use a json from one arc as a base for the next and generate a basefile automatically to insert the aliases into.

4. Engine changes
----

For the DLL this means to add another directory like current CG/CGAlt but for Ryukishi sprites. Unlike CGAlt this should not fallback to PS3 sprites - minor characters who don't have Ryukishi sprites simply won't have sprites instead of falling back to PS3.

Also we need a way for the user to configure which sprites they want to use. Since it's difficult for us to add a button to the config screen it would have to be another hotkey. Or possibly we can reuse the current hotkey for switching between sprites and just let it have a third value. There might be some compatibility issues with the config screen only having 2 values so it will need some testing.

5. Possible bugs
----

Our first patches had some issues with the engine freezing on some `WithFiltering` instructions if the file did not exist. Since then we made sure that no sprites are missing and it's likely the current engine won't actually freeze with all of our modding for LipSync. Still it needs some testing to check if it works ok since there will be missing Ryukishi sprite for the minor characters.

**Comment by drojf** [comment]
### Just a comment on 3. Write the list of aliases

For umineko I did something similar. I basically did the following:

- Compare the old and new scripts. "Line up" the two scripts like it would appear in winmerge. In Higurashi you can use the Japanese comment lines `//「俺にも判りそうなゲ...` to create aligned 'chunks' as below:

Old script
```
//「俺にも判りそうなゲームがあったらやらせてくれよ。＠
	OutputLine(NULL, "「俺にも判りそうなゲームがあったらやらせてくれよ。」",
		   NULL, "\"If there's a game you think I'd understand, let me play too.\"", Line_WaitForInput);
	OutputLineAll(NULL, "\n", Line_ContinueAfterTyping);


	DisableWindow();
	DrawBustshot( 1, "me_se_wa_a1", -160, 0, 0, FALSE, 0, 0, 0, 0, 0, 0, 0, 0, 200, TRUE );
```

New script
```
//「俺にも判りそうなゲームがあったらやらせてくれよ。＠
	if (GetGlobalFlag(GADVMode)) { OutputLine("<color=#956f6e>圭一</color>", NULL, "<color=#956f6e>Keiichi</color>", NULL, Line_ContinueAfterTyping); }
	ModPlayVoiceLS(3, 1, "ps3/s19/01/hr_kei00330", 256, TRUE);
	OutputLine(NULL, "「俺にも判りそうなゲームがあったらやらせてくれよ。」",
		   NULL, "\"If there's a game you think I'd understand, let me play too.\"", GetGlobalFlag(GLinemodeSp));
	if (GetGlobalFlag(GADVMode)) { ClearMessage(); } else { OutputLineAll(NULL, "\n", Line_ContinueAfterTyping); }


	DisableWindow();
	ModDrawCharacter(1, 3, "sprite/normal/me1a_warai_a1_", "1", -160, 0, 0, FALSE, 0, 0, 0, 0, 0, 0, 0, 0, 200, TRUE );
```

- Within each chunk, match up the drawBustShow and ModDrawCalls. If there is more than one on either side, just ignore it completely for now (keep it simple). This gives you a mapping as below:

`"me_se_wa_a1" ->"sprite/normal/me1a_warai_a1_"`

- Perform this mapping over as many scripts as possible. Record how many times a "ps3 sprite" maps to a "original sprite". You can also do the reverse mapping if this works better. Even if you get some bad matches, it doesn't matter, since you just take the "most popular" match out of all the matches you've seen.

```
"me_se_wa_a1" maps to: 
- 99 times "sprite/normal/me1a_warai_a1_"
- 13 times "sprite/normal/me1a_def_a1_",
- 5 time "sprite/normal/me1b_akuwarai_a1_"
```

In the above case, you can be pretty sure it matches `me1a_warai_a1_`


- Identify and poor or bad matches, And review them manually. For example:

```
"me_se_wa_a1" maps to: 
- 9 times "sprite/normal/me1a_warai_a1_"
- 1 times "sprite/normal/me1b_akuwarai_a1_"
```

Even though 90% of the time it matches `me1a_warai_a1_`, there are only 10 matches, so the confidence level is low that this is correct


```
"me_se_wa_a1" maps to: 
- 99 times "sprite/normal/me1a_warai_a1_"
- 96 times "sprite/normal/me1b_akuwarai_a1_"
```

It's not certain which one matches in this case, since they match a similar number of times.

- Output the mapping as an image, with the old sprite on the left, and new sprite on the right, 
for comparison (except the bad matches). This lets you qucikly check each sprite actually does 
match the each other visually.

#### Other notes:

- In umineko, the names were similar enough that you could sometimes match by name alone,
  or use it to complement other matching methods. Not sure if that could be done here.

- You can also sort the results by "total matches" and "most popular match / total matches", then look at the worst in each category, and also just analyze the numbers overall to see how good the matching is.

**Comment by drojf** [comment]
I wrote a test program to do the above matching (from ps3 to steam). Does this matching look reasonable? (I did it using all scripts in onikakushi which had the same name)

https://gist.github.com/drojf/57abc44ae6e2a615aca1818596aafe43

**Comment by enumag** [comment]
@drojf This will work well for sprites that are used many times but I'm not sure about sprites that are used rarely or even just once. Also the comments with japanese text are not entirely reliable - there are several known cases there these comments are not placed correctly in the script and are like 10-20 lines ahead or behind.

**Comment by drojf** [comment]
I'll try running it across all the game scripts and see what kind of results we get.

At the very least it should save you some time matching the 'common' sprites.

I might need to take into account the `sprite/night` and `sprite/sunset` etc versions. Is the PS3 name usually unique? EG for `'sprite/night/me2_akuwarai_a1_': {'me_si_aw_a1': 1},` if I just take the `me2_akuwarai_a1_` part, would it be unique?

**Comment by drojf** [comment]
I ran the program on himatsubushi/onikakushi/watanagashi/mekakshi, results are below

[results_2019-02-23.zip.zip](https://github.com/07th-mod/higurashi-patch-compiler/files/2896166/results_2019-02-23.zip.zip)

There are two sets of results - one where I ignore the sunset/normal/night, and one where I don't ignore it.

Included in each set of results is the ps3->steam mapping and steam->ps3 mapping.

You can open the .csv files in Excel or simlar. The files are sorted by number of matches. The confidence percentage column is just (highest number of matches / total number of matches * 100%).

The JSON .txt files has the raw match data, and is not sorted.

I don't think this is usable by itself, but I think it will be useful as a base for the alias list. The matches near the top of the list with high confidence can basically be used as is, but as you go further down the list you'd want to check those manually since they have few matches.

I added all the files to the repo (including the input scripts), so if anyone wants to run it, just clone the repo and run the .batch file: https://github.com/drojf/match_higurashi_scripts

**Comment by drojf** [comment]
On discord you requested that the program list all files, even files which weren't matched, so that the list of sprites was 'complete'. I made the change a while ago but never saved the output files, until just now.

The file where the sprite was last seen is is also added to the csv file, so it's easier to find if you want to check a particular sprite.

You can get the files in the `normal` and `ignore time of day` folders in the repo here: https://github.com/drojf/match_higurashi_scripts

**Comment by enumag** [comment]
The most useful result for us is most likely [ignore time of day, no reverse](https://github.com/drojf/match_higurashi_scripts/blob/master/ignore_time_of_day/ignore_time_of_day.txt.csv).

However it contains records for characters who don't have Ryukishi sprites such as Oryou, Kimiyoshi or Miyuki. Can you try to whitelist just the PS3 sprites with the prefixes of characters who do have Ryukishi sprites?

Here are the prefixes we need for sure but I might have missed some:

```
aka // Akane
aks# // Akasaka
ha# // Hanyuu
iri# // Irie
kasa // Kasai
kei# // Keiichi
me# // Mion
oisi# // Oishi
oko // Okonogi
re# // Rena
ri# // Rika
rina // Rina
sa# // Satoko
sato# // Satoshi
si# // Shion
ta# // Takano
tie // Chie
tetu // Teppei
tomi# // Tomitake
```

`#` means number. You can't just use prefix without number because `tomi` would include `tomita` who should be excluded. You can't easily add an underscore after because Hanyuu has sprites with prefixes like `ha3b_`. I recommend using a regexp for this.

----

I'm unsure if we should use Ryukishi sprites for child Mion (`chibimion` prefix) and little Rika (`rim` prefix). What do you think @idealpersona @ItaloKnox?

**Comment by drojf** [comment]
Added https://github.com/drojf/match_higurashi_scripts/blob/master/ignore_time_of_day/ignore_time_of_day.filtered.txt.csv

Didn't do any sanity checks though, might be worth checking everything as it should be.

I also fixed the output so it's deterministic, by sorting on the other columns instead of just the confidence column. So if you notice the rows have moved around, that's why. Hopefully the rows won't move around anymore.

**Comment by enumag** [comment]
Perhaps we should also drop the sprite / portrait distinction. They are different size but the same facial expression. What do you think?

**Comment by drojf** [comment]
Hi enumag,

I've gone ahead and dropped the distinction with the portrait sprites. This means the matching portion treats all the variants (portrait, night, etc) the same when it sees them in the game script. 
 
I also was preparing a tool so that @orian34 could do the manual matching part easier. While I was creating that tool, and going through it with orian, we found the following:

- The 'Chibi' PS3 sprites don't exist in the ryukishi sprites (orian mentioned you created chibi MG sprites in the past?). Eg `sprite/flashback/chibimion_def_0.png`
- There are some characters which don't exist at all in the ryukishi sprites, like `tomita`
- For the PS3 Sprites, the variants (portrait, sunset, night etc.) are not always a subset of the "normal" images. There are a couple images which only exist as their portrait/sunset etc. variants. I think this is not a problem, but just something to be aware of (or it indicates I've messed up copying files). I can make a list of these sprites if necessary. For example, there does not appear to be a 'normal' variant of `sprite/sunset/iri2_warai_<0-2>.png`, only the sunset variant exists.

None of these stops orian fixing up where the auto-matcher failed, so we'll go ahead and do that regardless. Once that is finished, we'll post the matching here.

For the missing sprites, let us know what the plan is. For now we'll just mark them in the mapping so it's obvious which ones need to be fixed/created.

**Comment by drojf** [comment]
Orian has gone through the images now. I've gone through the ones which are still not matched and produced a zip file with each non-matching image categorized

- CHIBI_SPRITE: There are no chibi ryukishi sprites, so these could not be matched.
- NEW_CHAR_[charname]: A character who has a sprite in the ps3 version, but never had a ryukishi sprite
- NO_MATCH/NO_SPRITE: The character in ps3 and ryukishi, but that particular facial expression does not exist. Otherwise, a catch all category for no-match.
- RENA WITH SWORD: Sprites with rena holding a sword

[unmatched_sprites.zip](https://github.com/07th-mod/higurashi-patch-compiler/files/3269096/unmatched_sprites.zip) (total: 81 unmatched)

For the "NO_MATCH/NO_SPRITE", there is probably a very similar sprite we can use, even if it's not exactly the same one. If that's OK we can probably fill some of those ones in (there are 18 of that type of match failure)

The updated mapping is here: https://github.com/drojf/match_higurashi_scripts/blob/master/imageComparer/noconsole_output.txt.csv

**Comment by enumag** [comment]
> The 'Chibi' PS3 sprites don't exist in the ryukishi sprites (orian mentioned you created chibi MG sprites in the past?). Eg `sprite/flashback/chibimion_def_0.png`

Yeah, these were custom created by @jwgrlrrajn for both Rika and Mion. I reckon we should treat them the same as the nonexistent characters - ignore them.

> For the PS3 Sprites, the variants (portrait, sunset, night etc.) are not always a subset of the "normal" images. There are a couple images which only exist as their portrait/sunset etc. variants. I think this is not a problem, but just something to be aware of (or it indicates I've messed up copying files). I can make a list of these sprites if necessary. For example, there does not appear to be a 'normal' variant of sprite/sunset/iri2_warai_<0-2>.png, only the sunset variant exists.

This is completely ok. We have the normal variants in the full sprite packs. The individual patches only contain sprites that are actually used in the game. If some sprite is only used at night, then the patch only contains the night variant. Same goes for portraits too.

> For the missing sprites, let us know what the plan is. For now we'll just mark them in the mapping so it's obvious which ones need to be fixed/created.

We simply won't show any sprites. I'm not exactly sure how to do this in the game scripts though. Like do we need to put some condition around the call? Previously some draw functions crashed when the image did not exist but those cases might be fixed now. Anyway I's like to have @idealpersona's recommendation on this.

> For the "NO_MATCH/NO_SPRITE", there is probably a very similar sprite we can use, even if it's not exactly the same one. If that's OK we can probably fill some of those ones in (there are 18 of that type of match failure)

Yeah this is similar to the CGAlt thing. MG sprites also have less facial expressions than console releases so some sprites are duplicated as a result. Match the closest sprite you find.

> edit: Orian brought up that a couple console sprites are not in there....that is because when I was writing my program I'm like "there's nothing to match against for the console arcs, so I'll just ignore them", but if we do want to support switching to ryukishi sprites for the console arcs, we'll need to also match any sprites that appear there.

console arcs have so many exclusive characters that it's not worth it adding ryukishi sprites support there in my opinion

**Comment by enumag** [comment]
Also what is the plan on how we treat this? Are the ryukishi sprites like an installation option? If so do they replace CG or CGAlt? Or are we adding a third CG directory and changing the DLL to work correctly?

**Comment by jwgrlrrajn** [comment]
For chibi Rika and Mion, the original just had them show up as their normal ages. I suppose we could shrink the Ryukishi07 sprites and modify them like we did the MG, but that may be more trouble than it's worth. Alternately, we just have them show up as-is.

**Comment by ItaloKnox** [comment]
@enumag I would prefer if it was part of the default installation. We could just rename the original ``CG`` folder to something like ``OG`` and leave it like that. We could also go even further and include upscaled sprites and backgrounds in the installer as an optional download.

**Comment by enumag** [comment]
@ItaloKnox I don't think I understand your thoughts. Let's talk on Discord tomorrow.

**Comment by enumag** [comment]
@drojf recap of our dicussion on Discord:

I'll be using this file to generate the graphics packs: https://github.com/drojf/match_higurashi_scripts/blob/master/imageComparer/noconsole_output.txt.csv

For now it doesn't include Minagoroshi so please add that (this is the higher priority task).
Original game scripts: https://cdn.discordapp.com/attachments/430265398810640394/642844630588850226/Scripts.7z
Patched scripts: https://github.com/07th-mod/minagoroshi/tree/master/Update

Secondly we need matching for backgrounds too. Try to do that the same way as sprites to see if we can continue from there.

That's all, thanks!

**Comment by enumag** [comment]
@drojf another issue discovered with the linked CSV: child Mion and Rika are missing here. They shouldn't, in this case we want to use their standard sprites. Resolving this has the same priority as Hanyuu.

**Comment by enumag** [comment]
@drojf And another issue for tsumi: there is no matching for `renasen` sprites (battle-mode Rena with her hatchet). Again we want to use normal Rena sprites here.

**Comment by drojf** [comment]
I just fixed the chibi sprites and rena with sword sprites in https://github.com/drojf/match_higurashi_scripts/commit/582af9d72061cc46b5df914304fd93a6e3f59ce5 (note: I also removed the last line of the file which said "IGNORE NOT A SPRITE")

I added the missing lines from minagoroshi in https://github.com/drojf/match_higurashi_scripts/commit/f46cfe991ff7ea80d671277a535b66f3a0a8026b

I'm sure there'll be more things but I think that's everything you asked for so far (except for the backgrounds)

**Comment by enumag** [comment]
@drojf Thanks, that seems to be everything, couldn't find any further issues. Focus on the backgrounds now. ;-)

**Comment by drojf** [comment]
I did the background matching just now, results below. These results have been automatically matched, then I flicked through them with my comparison tool manually to check they actually were correct, and manually fixed any missing matches.

https://github.com/drojf/match_higurashi_scripts/blob/master/background_matching/background_matching_ignore_variants.csv

Rows with "-1" as the match count have been manually fixed by me. There should be no rows which failed to match.

Some things to note:
- only images in the background folder are processed
- blur and flashback type images are treated the same as normal images (ignored if they have one of the same name, basically)
- while I still have the code which checks for any "missed" images, you may want to double check if I missed processing any images - I might have broken that part while I was making the backgrounds work, not sure
- You'll have to generate the blur and flashback type images yourself (I assume you've already done that for the sprites)

**Comment by drojf** [comment]
I just noticed that Naegles (are they on Github?) already did the background matching manually for the question arcs. If you're manually checking the backgrounds, this would be a place to start. 

They also said they were doing the answer arcs, but I'm not sure what the status is - I'll try to ask them

[Higurashi_Background_Match.xlsx](https://github.com/07th-mod/higurashi-patch-compiler/files/5139809/Higurashi_Background_Match.xlsx)

**Comment by Naegles** [comment]
The answer arcs should be much easier to do tbh.
The names actually sorta match up for that.
Also, while I'm pretty certain I got most stuff correct someone might still want to check my work on the question arcs before using it.
There was some stuff I wasn't 100% sure about.

**Comment by drojf** [comment]
I had a short discussion with Naegles on Discord:
- I need to re-run the automatic background matching to add matsurabayashi's new backgrounds
- I would like to then compare Naegles manual results and my automatic results for the question arcs, and see which ones don't match. Then we can fix those ones up.
- After that Naegles mentioned they wouldn't mind checking over my automatic matches for the Answer arcs by using the comparer tool

I'll try to follow this plan and hopefully we can get the matching done, then it's just the upscaling/whatever post processing we want to apply that needs to be done.

**Comment by drojf** [comment]
So here's what I ended up doing

- I used Naegle's matches as the 'base' matching
- Then, if there were any images not in naegle's matching, I filled them in with my automatic matching

Last weekend Naegles went through all the matches with the comparer tool and checked them, correcting about 70/400 matches.

I've produced a new file with these matches (tidied up a bit). 

Background Matching CSV File:
https://github.com/drojf/match_higurashi_scripts/blob/master/background_matching/manual_bg_map_paths_generated.csv

There are a few things to take note of:
- There are now more/different columns in the file. The ones you'll need are:
  - "ps3_name" (ind 1): The PS3 filename of the file.
  - "final_choice" (ind 4): The ryukishi image filename which best corresponds to the ps3 file.
    - This will be set to "NO_MATCH" if there is no match for this file
    - Note that this is now index 4 rather than 2 as it was for the sprites .csv
  - "ps3_path" and "ryukishi_path" (ind 6 and 7): The path to the ps3 and ryukishi images.
    - This path won't match the folder structure of the installed game - you will need to substitute `ep1` with `Higurashi When They Cry\HigurashiEp01_Data\StreamingAssets\CG` for example.
- There are both CG images and backgrounds included in this file. To determine what kind of image it is, check the "ps3_path" column and look for `\background\` for backgrounds etc. Basically check if the file is in the "backgrounds" folder using the included path.
  - For now, I think we should only consider images in the `background` folder of each episode.
  - All images in the "background" folder should have a match in this file (none of them should be NO_MATCH)
- As before with the sprite matching, variant images (like those in the `background\blur` or `background\flashback` folders) will need to be generated from their base images (the variants do not appear in the csv file).
- I have checked and there are no name conflicts within the same episode (as in, two files with the same filename in the same episode, but with different paths)

I can do some more pre-processing on the csv file if it makes it easier for you to generate the images

**Comment by drojf** [comment]
@enumag are you still up for converting the backgrounds using the mapping etc. like you did the sprites? I think I got everything ready but if not let me know.

**Comment by ItaloKnox** [comment]
Which chapters would that cover? All of them? Let me know before you start since I have a ton of translation changes to review and that would be the perfect time to publish them.

**Comment by enumag** [comment]
@drojf I'd like to do it but sadly I'm gonna be very busy for the next month or so. 😞

**Comment by drojf** [comment]
@enumag Would it be possible for me to look over/run your php code to generate the images? Or if it's simple enough maybe I can just...do it in Python?

From my understanding, I just need to apply the correct effect to the images in the subfolders (like the `flashback` folder is sepia, `blur` folder is blurred, `greyscale` has greyscale applied). And all other images (for now) will just be stretched to 16:9. But there might be some other things which I'm not aware of. I would probably want to refer to your code for the exact filters you used so I can make sure they are consistent (like, in case I apply a slightly different sepia function).

If I can do this image conversion part, then at least that's out of the way....although I'm not sure what is involved after that (if there need to be script changes, or if it's just a single .dll change to enable toggling backgrounds) @idealpersona / @tellowkrinkle is the background toggling already implemented in the .dll?

@ItaloKnox this should cover all the chapters (well, except console). I'm not sure when it will be ready though....I'm a little confused by "Let me know before you start", because I've already....started? Do you mean before it's finished, so we can publish it all together?

**Comment by enumag** [comment]
IIrc all background image effects were made by hand, not automated. But maybe it would be okay for these?

Feel free to use my code ofc. Unfortunately I don't have time to adapt it myself now.

**Comment by ItaloKnox** [comment]
> @ItaloKnox this should cover all the chapters (well, except console). I'm not sure when it will be ready though....I'm a little confused by "Let me know before you start", because I've already....started? Do you mean before it's finished, so we can publish it all together?

Yeah, that's what I wanted to say. Just let me know before everything is done because we might have more translation PRs to check. I already looked at everything from chapters 1 to 4, but didn't merge them yet. This update implies rebuilding the graphics pack, so we can't leave any images behind from the PRs.

**Comment by drojf** [comment]
So, I think I got most images copied/stretched out (it's mostly automated so I can run it again if need be). The following files are replaced:

- Everything in the root of the CG folder `CG/*.*` (this folder contains splash images, menu images and menu backgrounds, and effect masks)
- Everything in the `CG/background` folder 
    - subfolders with greyscale, negative, and blur images have the correct effect applied (I had to do the zoom blur images manually in GIMP, but those are only used for one chapter and I've saved them for future runs)
- Everything in the `CG/flashback` folder
    - this folder only exists for eps1-4 - it contains sepia versions of existing images
- Everything in the `CG/scenario` folder 
    - only the `background.png` image is actually changed, but I copy all images for consistency. 
    - The `background.png` is always set to the ryukishi `scenario_a.png` image for that episode

From what I understand, I should create a new folder like `OGBackgrounds`, containing files which the game should "prefer" to the original images. Then the game engine will prefer those images when the option is enabled, and fall back to the `CG` folder if there is no such image. Is that how it will work?

I have resized all images to 1080p for now. If the engine can resize the images, you could use much smaller images (853x480) and let the engine resize them, saving lots of space. I think the quality of the resize algorithm doesn't really matter since the images are blurry anyway.

- [ ]  One thing we should remember to do before release is to optimize all the .png images.

I've attached all the images, but at a very small resolution, just so you can get some idea of the folder structure. Each `epX` folder contains the CG folder images for that episode. Download it at this OneDrive link: https://1drv.ms/u/s!Ar-lAVeetlqhgsY6iO0hncY6LJU5xA?e=EM2kJo

**Comment by drojf** [comment]
Ok, I just read https://github.com/07th-mod/higurashi-assembly/issues/33 which is where Tellow implemented ryukishi sprites feature (or rather the generic ability to swap art sets).

After following the instructions to edit the `init.txt` file for onikakushi from

`ModAddArtset("Original", "オリジナル", "OGSprites:CG");`

to

`ModAddArtset("Original", "オリジナル", "OGBackgrounds:OGSprites:CG");`

I can successfully get the backgrounds to toggle when you press "P" (along with the sprites).

I guess some people may want other combinations (like the current ryukishi sprites with PS3 backgrounds). Not sure if you guys want to support that by adding an additional artset command.

I'll do some more testing, but I think we just need to produce the graphic packs and update the `init.txt` files for each chapter, besides and QA testing that needs to be done (could distribute it to the people who requested ryukishi backgrounds for initial testing).

**Comment by drojf** [comment]
I've gone through the sprite mapping for Rei. There aren't many new sprites - I've added 24 new rows to the matching file (rows 502-526 at the end): https://github.com/drojf/match_higurashi_scripts/blob/master/imageComparer/noconsole_output.txt.csv

Notes:
 - My program doesn't exclude portrait sprites, so they appear in the list, but should be fairly harmless (the new hanyu `ha4` sprites, rows 503-507 are actually just portrait sprites, see below)
 - Keichi in a maid outfit `sprite/kei6_komaru_` and `sprite/kei6_nayamu_` is not included in the ryukishi sprites at all, so I've written NEW_CHAR_KEI_MAID so they are excluded
 - Kameda has a new sprite, and they don't have a ryukishi sprite either, the same as previous chapters `sprite/kameda2b_odoroki_`


## Portrait Sprites

Portrait sprites will be added manually (there are only 7), I've prepared non-zoomed in images here: https://github.com/drojf/match_higurashi_scripts/tree/master/preprocessed_sprite/ep9/portrait . We can get someone to zoom them in later, or just leave them as-is.

## Generating sprite files

Back in 2019, I think @enumag generated the sprites for me from the mapping file. Then in 2020, I generated the backgrounds myself.

So I think we need to run enumag's script for the sprites, however enumag will need to change their script to acommodate the following:

 - **IMPORTANT**: for rei, the folder structure for the unmodded game sprites has changed - sprites are now sorted into subfolders, with the 'default' outfit in a folder like `sprites/okonogi/oko_ikari.png` and variants in a deeper subfolder like `sprites/rika/nekomimi/ri_defa1.png`
 - **IMPORTANT**: One new column "manual match" has been added to the csv file, to be consistent with the background matching .csv. It shouldn't affect your code though.
- Just as a reminder, if the match says `NEW_CHAR_` + something, it means that this is a PS3 exclusive character, so no ryukishi sprite exists, and no file should be outputted.
- Another reminder - my mapping only lists files which are actually used in the game script. As a result, we may have some files in our mod sprite folder which are not in this list.

Please use the updated sprite mapping here: https://github.com/drojf/match_higurashi_scripts/blob/master/imageComparer/noconsole_output.txt.csv
